pyomo/common/factory.py

```diff
@@ -0,0 +1,67 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and 
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain 
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+
+class Factory(object):
+    """
+    A class that is used to define a factory for objects.
+    """
+
+    def __init__(self, description=None):
+        self._description = description
+        self._cls = {}
+        self._doc = {}
+
+    def __call__(self, name, exception=False):
+        name = str(name)
+        if not name in self._cls:
+            if not exception:
+                return None
+            if self._description is None:
+                raise ValueError("Unknown factory object type: '%s'" % name)
+            raise ValueError("Unknown %s: '%s'" % (self._description, name))
+        return self._cls[name]()
+
+    def __iter__(self):
+        for name in self._cls:
+            yield name
+
+    def __contains__(self, name):
+        return str(name) in self._cls
+
+    def get_class(self, name):
+        return self._cls[name]
+
+    def doc(self, name):
+        return self._doc[name]
+
+    def unregister(self, name):
+        name = str(name)
+        if name in self._cls:
+            del self._cls[name]
+            del self._doc[name]
+    
+    def register(self, name, doc=None):
+        def fn(cls):
+            self._cls[name] = cls
+            self._doc[name] = doc
+            return cls
+        return fn
+
```

Comment:
You should implement the following for backwards compatibility:
  - `services(self)`
  - `activate(self, name)` [a wrapper around register? or better let, move it over to a`self._deactivated` dict for possible reactivation later]
  - `deactivate(self, name)` [the inverse of deactivate]

...all should issue deprecation warnings.

Reply:
I am not sure why you are introducing an unnecessary backwards incompatibility, though.  If we are keeping things like the context manager API because "old plugin system supported this, and developers started using this API", the same argument could/should be made here.  These methods are not fundamentally difficult to implement, and would not necessarily change the design of the new factories.

Reply:
But these factories don't support the same type of plug-in functionality.  So there's not motivation for backwards compatibility.  Backwards to what?

For example, I've intentionally removed the alias() and implements() function, and the class dependency on Plugin.  Are you suggesting that I reintroduce them? That would make sense if I was replacing the PyUtilib plugin system with a different plug-in system.

But that's not what's happening here.  We *were* using a very general plug-in system to create factories, which required a complex API to interface.  But *now* we're using a simpler API that is easier to understand and apply. 

Thus, adding the "backwards compatibility API stuff" both (a) confuses developers about the nature of the change, and (b) misses the point of creating a simpler API.

Reply:
`services()` returns the list of registered names.  That functionality still makes sense for the new plugins.  Similarly, `activate` and `deactivate` is functionality that could easily supported, and is almost analogous to register/unregister.   I think these methods are different from alias/implements because alias/implements only affects developers who are creating new plugins, whereas services in particular (and activate/deactivate to a lesser extent) are used by consumers of the plugin system.  Further, the implementation of those methods is completely independent from the metaclass-magic, whose removal was your motivation for the PR (whereas alias/implements were intimately tied to it).

Reply:
Note that services() has been replaced by __iter__().  The functionality hasn't been removed, but the API has been simplified.

I could add activate()/deactivate() as you suggest.  But to what end?  My changes removed *all* use of this logic.  Further, I have consistently struggled to parse what that logic means.  Hence, I'm inclined to follow the KISS principle.

(Maybe we'll find that the use of activate/deactivate was well-motivated.  @ghackebeil, many of the calls to activate/deactivate were in pyomo.pysp, so you might want to weigh in on this API change.)

Reply:
Yes - I realize that services/activate/deactivate now have new preferred APIs, and that you have removed all references in the core.  The point is to provide backwards compatibility for *user* code (that we don't / can't update).  This is why I advocated implementing them as `@deprecated` methods: it makes it trivial to provide users with informative messages as to how to update their code *without breaking their code*.

Personally, I think that the `__enter__`/`__exit__` methods should also be deprecated.

Reply:
Since we're changing the API altogether, I added exceptions when these methods are called rather than deprecating them.

Note that the __enter__/__exit__ methods are not on the factory objects.  I'm loath to deprecate them without reviewing why developers thought they were convenient.  For example, it might be semantically more clear to manage software licensing with these methods, which seem analogous to freeing up resources with file I/O context methods.